### PR TITLE
Make sure that *.Linear modules only export linear functions

### PR DESCRIPTION
**Pull request by utdemir**
I created this PR after my comment on #127:

> When I import a *.Linear module, I expect to only get linear functions. and the data types those functions work on. It is surprising to me that I can import Prelude.Linear and it contains non-linear functions.
> In some cases using *.Linear modules require some data types from base to be imported (eg. Data.Maybe.Linear does not export Maybe (..), or Data.Ord.Linear does not export Ordering), which is a bit annoying.

The main changes are

* `Prelude` re-export from `Prelude.Linear` is removed. Instead I only exported common data types like `Int` and `Bool`.
* Now `Data.Maybe.Linear`, `Data.Either.Linear` and `Data.Ord.Linear` exports `Maybe(..)`, `Either(..)` and `Ordering(..)` respectively.
* `Data.Semigroup.Linear` used to export non-linear destructors for newtype wrappers like `Sum`. I implemented linear versions of them manually.

The rest is pretty much just fixing the import lists and qualifying some names. I tried to not make any controversial changes, so let me know if you prefer anything different. However, this PR does cause the API to change quite a bit, sometimes in unexpected ways (eg. now `Prelude.Linear.Eq` is `Data.Eq.Linear.Eq` instead of `Prelude.Eq`); but I do not think we guarantee backwards compatibility yet.

### CODE REVIEW

**Comment by aspiwack** [comment]
A few comments on spirit of all this, before I review:

- `Prelude.Linear` is meant to be a replacement for `Prelude`. More or less drop-in. So we should have a similar content. It's true that `Prelude` probably exports too many things. But we should make conscious decisions for what goes in and what goes out. That was the rationale for re-exporting everything and hiding things I didn't want. But a more principled approach would be to have a list of what is in `Prelude` and not in `Prelude.Linear`, and decide what to exclude and what to port.
-  It's alright for a non-linear function to be exported by a linear module. There is at least one (the constructor for `Unrestricted` (or whatever the name will end up being as per #131 ). When a linear variant of the function makes sense, it should be the exported function. But a function not having a linear arrow in it is not ground for exclusion.

**Comment by utdemir** [comment]
My reasoning is only based on my experience developing on a few small linear modules, so not a very informed opinion; but here is what I think:

1. With the current state of the library, we can not use linear-functions as non-linear functions. This might change if we were to write all our functions in a multiplicty-polymorphic way.
2. An average application uses both linear functions and non-linear functions. This can be observed on most of our medium-sized modules on `linear-base` where we import both `Prelude` and `Prelude.Linear`. And I might actually even claim that in an average application there'll be much more non-linear functions than linear functions.

So, for now the reader needs to be aware of which functions are linear and which are not. One way to make it easier is importing `Prelude.Linear` qualified, then the `Linear.` (or `L.`) prefix clearly distinguishes linear functions. 

> Prelude.Linear is meant to be a replacement for Prelude. 

I think before most of our functions are linearity polymorphic, we can not reasonably expect people to not use `Prelude` at all. 

---

So, in short; I'd agree with re-exporting non-linear functions from linear-base once we're sure that our `Prelude.Linear` can provide the same functionality as `linear-base`. But even in that case, I'd be happier with an explicit export list.

**Comment by mboes** [comment]
> With the current state of the library, we can not use linear-functions as non-linear functions. This might change if we were to write all our functions in a multiplicty-polymorphic way.

This is only true of higher-order functions, right?

**Comment by utdemir** [comment]
> This is only true of higher-order functions, right?

To be honest, I did not know that, but it seems so :). In my defense, I was misled by the existence of `forget :: (a #-> b) #-> a -> b` function.

My point is now a bit weaker, but still stands since we are also exporting higher order functions.

**Comment by aspiwack** [comment]
Mostly the intent is that you would import `Prelude.Linear` unqualified, and qualify base's `Prelude` when needed (though it may be the opposite in a module with very little linear types I suppose). But yes, things will be better when we can make things polymorphic.

But it doesn't deter the argument: linear modules export functions which are relevant to the stuff at hand. The `Unrestricted` data type comes with a decidedly unrestricted function: `Unrestricted :: a -> Unrestricted a`, as I was saying. Any module exporting a record will come with its projections, which are unrestricted as well. You shouldn't have to import a second module to import a function just because it doesn't have linear arrows in it.

However, the spirit is that _when a function can have linear arrows_ it should. In linear-base, I mean.

**Comment by utdemir** [comment]
Okay. Then I'll modify this PR by re-adding the `Prelude` re-export but with an explicit export list, and while doing that I can note the functions we should implement with linear arrows.

**Comment by utdemir** [comment]
We decided that we want to implement something like this after the other PR's are cleaned up, and with using small PR's with easier discussions. So I am closing this one.
